Sources/Apollo/InMemoryNormalizedCache.swift

```diff
@@ -1,21 +1,37 @@
+import Dispatch
+
 public final class InMemoryNormalizedCache: NormalizedCache {
   private var records: RecordSet
-
+  private let queue: DispatchQueue
+  
   public init(records: RecordSet = RecordSet()) {
     self.records = records
+    self.queue = DispatchQueue(label: "com.apollographql.InMemoryNormalizedCache", attributes: .concurrent)
```

Comment:
Shouldn't we be using a serial queue to ensure that we don't have two blocks accessing it at the same time? Wouldn't concurrent still allow for race conditions?

Reply:
I was thinking same, but serial queue would slow down things and I can't replicate crash with concurrent one... Maybe fact that it is accessed on same thread is enough...
We have this PR in production for past couple of months (way before I created PR) and see 0 crashes.

Reply:
Okay, you're right a serial queue is less performant, but this concurrent queue doesn't seem to necessarily address the issue. Of course it's up to @designatednerd in the end, and I'm glad that this workaround has been able to work for your team in production. But I wouldn't feel comfortable about pulling in a workaround for an issue that we can't consistently reproduce and that doesn't seem to make perfect sense.

It seems that this may _happen_ to avoid the race conditions present in your code base currently, but it doesn't completely avoid the possibility of them occurring. 

I'd feel much more comfortable going with `NSRecursiveLock` as a solution here than dispatching to another queue.

Additionally, though I understand it may not be possible, it would be great if we could work a bit more on trying to create a test case that consistently crashes.

I'd be willing to spend some time looking into this and try to reproduce a crashing test, but not sure I'll get much further than you did @RolandasRazma .

Reply:
completely agree it needs better testing. But the only thing I can think off is add check if ivar is accessed from multiple threads

Reply:
Hello, I've been looking around to see all the work being done on this library recently and saw this PR and this thread, so hope you don't mind me chiming in. The reason why concurrent works is because your using it with barriers. If you only even access a concurrent queue with barriers, you've essentially made it function exactly like a serial queue. A common approach to synchronizing with a queue, is using a concurrent queue, then use async barrier writes, and sync non-barrier reads. Keeps the happy-path fast, and the write path safe.

Reply:
IIRC, I think async reads would work with the pattern described too, but would definitely need testing because I'm not 100% sure of any side-effects inside the task blocks here... if any. Otherwise, might as well take the first suggestion, if it's all "barrier'ed up", might as well make it serial and remove the barriers.

Reply:
Yeah, if you are always using barriers, then serial is the same! Thanks @jimissacs!